inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py

```diff
@@ -98,11 +102,19 @@ def __init__(self) -> None:
 
     def add(self, parameters: ApplicationAddSourceParameters) -> None:
         """Adds a source file and optional GPG key to be used during Ubuntu application updates."""
-        # Step 1: Add key (Optional)
+        # Step 1: Verify gpg key uri from trusted repo list 
         if parameters.gpg_key_name and parameters.gpg_key_uri:
+            try:
+               url = parameters.gpg_key_uri
+               #URL slicing to remove the last segment (filename) from the URL 
+               source = url.value[:-(len(url.value.split('/')[-1]) + 1)]
```

Comment:
**Suggestion:** The 'source' variable slicing logic is not clear and might be error-prone. Consider using 'urllib.parse' to handle URL parsing more robustly. [enhancement]
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
               from urllib.parse import urljoin, urlparse

               url = parameters.gpg_key_uri
               # Use 'urllib.parse' to handle URL parsing
               source = urljoin(url.value, urlparse(url.value).path.rpartition('/')[0] + '/')
               verify_source(source=source, dispatcher_broker=Dispatcher_Broker)
```